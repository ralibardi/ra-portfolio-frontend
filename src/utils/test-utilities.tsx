import { ThemeProvider } from '@contexts/theme-context';
import {
  act,
  type RenderHookOptions,
  type RenderHookResult,
  type RenderOptions,
  type RenderResult,
  render,
  renderHook,
  screen,
  waitFor,
} from '@testing-library/react';
import type { FunctionComponent, ReactElement, ReactNode } from 'react';
import { MemoryRouter as Router } from 'react-router-dom';

interface IAllProvidersProps {
  children: ReactNode;
}

// Mock ToastProvider for testing
const MockToastProvider: FunctionComponent<{ children: ReactNode }> = ({ children }) => {
  return <>{children}</>;
};

const AllProviders: FunctionComponent<IAllProvidersProps> = ({ children }) => (
  <Router>
    <ThemeProvider>
      <MockToastProvider>{children}</MockToastProvider>
    </ThemeProvider>
  </Router>
);

const customRender = (ui: ReactElement, options?: RenderOptions): RenderResult =>
  render(ui, { wrapper: AllProviders, ...options });

const customRenderHook = <TProps extends IAllProvidersProps, TResult>(
  callback: (props: TProps) => TResult,
  options?: RenderHookOptions<TProps>,
): RenderHookResult<TResult, TProps> =>
  renderHook(callback, {
    wrapper: (props) => <AllProviders {...props} />,
    ...options,
  });

export { customRender, customRenderHook, act, screen, waitFor };
