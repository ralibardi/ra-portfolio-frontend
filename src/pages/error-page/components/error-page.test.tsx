import React from 'react';
import { act, customRender, screen } from '@utils/test-utilities';
import ErrorPage from './error-page';

describe('ErrorPage', () => {
  it('renders the error message', async () => {
    customRender(<ErrorPage />);

    const { headingElement } = await act(() => {
      const headingElement = screen.getByTestId('error-heading');
      return { headingElement };
    });

    expect(headingElement).toBeInTheDocument();
  });

  it('renders the correct heading', async () => {
    const error = new Error('Test error');
    customRender(<ErrorPage error={error} />);

    const { headingElement } = await act(() => {
      const headingElement = screen.getByTestId('error-heading');
      return { headingElement };
    });

    expect(headingElement).toHaveTextContent('Oops... Something went wrong');
  });

  it('renders the correct paragraph', async () => {
    const error = new Error('Test error');
    customRender(<ErrorPage error={error} />);

    const { paragraphElement } = await act(() => {
      const paragraphElement = screen.getByTestId('error-details-message');
      return { paragraphElement };
    });

    expect(paragraphElement).toBeInTheDocument();
  });

  // Additional tests to increase coverage

  it('does not render error details when no error prop is provided', () => {
    customRender(<ErrorPage />);
    expect(screen.queryByTestId('error-details')).not.toBeInTheDocument();
  });

  it('renders error details with correct content when error prop is provided', async () => {
    const error = new Error('Test error');
    error.name = 'TestError';
    error.stack = 'Test stack trace';
    customRender(<ErrorPage error={error} />);

    const { detailsElement, summaryElement, stackElement } = await act(() => {
      const detailsElement = screen.getByTestId('error-details');
      const summaryElement = screen.getByTestId('error-details-message');
      const stackElement = screen.getByTestId('error-details-stack');
      return { detailsElement, summaryElement, stackElement };
    });

    expect(detailsElement).toBeInTheDocument();
    expect(summaryElement).toHaveTextContent(`${error.name} - ${error.message}`);
    expect(stackElement).toHaveTextContent('Test stack trace');
  });

  it('calls window.history.back when the back button is clicked', async () => {
    const backSpy = jest.spyOn(window.history, 'back').mockImplementation(() => {});
    customRender(<ErrorPage />);
    const { backButton } = await act(() => {
      const backButton = screen.getByRole('button');
      return { backButton };
    });
    act(() => {
      backButton.click();
    });
    expect(backSpy).toHaveBeenCalled();
    backSpy.mockRestore();
  });

  it('matches snapshot when error is provided', () => {
    const error = new Error('Snapshot error');
    error.name = 'SnapshotError';
    error.stack = 'Snapshot stack trace';
    const { container } = customRender(<ErrorPage error={error} />);
    expect(container).toMatchSnapshot();
  });
});
