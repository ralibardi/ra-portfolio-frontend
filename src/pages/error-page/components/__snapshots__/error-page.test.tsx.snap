// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorPage matches snapshot when error is provided 1`] = `
<div>
  <div
    class="container"
    data-testid="error-page-container"
  >
    <div
      class="header"
      data-testid="error-header-container"
    >
      <button
        class="button"
        data-testid="button-with-icon-container"
        id=":r6:"
        type="button"
      >
        <div
          class="content"
          data-testid="button-with-icon-label-container"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-arrow-left icon"
            data-icon="arrow-left"
            data-prefix="fas"
            data-testid="button-with-icon-icon"
            focusable="false"
            role="img"
            viewBox="0 0 448 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M9.4 233.4c-12.5 12.5-12.5 32.8 0 45.3l160 160c12.5 12.5 32.8 12.5 45.3 0s12.5-32.8 0-45.3L109.2 288 416 288c17.7 0 32-14.3 32-32s-14.3-32-32-32l-306.7 0L214.6 118.6c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0l-160 160z"
              fill="currentColor"
            />
          </svg>
        </div>
      </button>
      <h1
        class="heading"
        data-testid="error-heading"
      >
        Oops... Something went wrong
      </h1>
    </div>
    <details
      class="description"
      data-testid="error-details"
    >
      <summary
        class="message"
        data-testid="error-details-message"
      >
        SnapshotError
         - 
        Snapshot error
      </summary>
      <pre
        class="stack"
        data-testid="error-details-stack"
      >
        Snapshot stack trace
      </pre>
    </details>
  </div>
</div>
`;
